src/coreclr/nativeaot/Runtime/MathHelpers.cpp

```diff
@@ -18,8 +18,46 @@ FORCEINLINE int64_t FastDbl2Lng(double val)
 #endif
 }
 
+//------------------------------------------------------------------------
+// TruncateDouble: helper function to truncate double 
+//                 numbers to nearest integer (round towards zero).
+//
+// Arguments:
+//    val  - double number to be truncated.
+//
+// Return Value:
+//    double: truncated number (rounded towards zero)
+// 
+double TruncateDouble(double val)
+{
+    int64_t *dintVal = (int64_t *)&val;
+
+    uint64_t uintVal = (uint64_t)*dintVal;
+    int exponent = (int)((uintVal >> 52) & 0x7FF);
+    if (exponent < 1023)
+    {
+        uintVal = uintVal & 0x8000000000000000ull;
+    }
+    else if (exponent < 1075)
+    {
+        uintVal = uintVal &  (unsigned long long)(~(0xFFFFFFFFFFFFF >> (exponent - 1023)));
+    }
+    int64_t intVal = (int64_t)uintVal;
+    double *doubleVal = (double *)&intVal;
+    double retVal = *doubleVal;
+
+    return retVal;
+}
+
 EXTERN_C NATIVEAOT_API uint64_t REDHAWK_CALLCONV RhpDbl2ULng(double val)
 {
+#if defined(TARGET_X86) || defined(TARGET_AMD64)
+
+    const double uint64_max_plus_1 = -2.0 * (double)0xFFFFFFFF;
+    val = TruncateDouble(val);
+    return ((val != val) || (val < 0) || (val >= uint64_max_plus_1)) ? 0xFFFFFFFF : (uint64_t)val;
+
+#else
```

Comment:
@tannergooding then do you think we should make this change?
I am referring to https://github.com/dotnet/runtime/pull/84384#issuecomment-1604693809 . Since it was suggested we make apt changes in nativeAOT and mono. From https://github.com/dotnet/runtime/pull/84384#discussion_r1240569897, I understand that the behavior can be different and user can decide what to use.

@jkotas do you agree?

Reply:
The behavior of floating to integer conversion is speced as undefined.  We know from past experience that applications and libraries out there depend on the exact behavior even though it is undefined. Changing the exact behavior is going to break some of them. 

It would be best to do one clean documented breaking change to get us to the defined saturating behavior that we have agreed on. It may be too late to get this breaking change into .NET 8.

Reply:
I think that makes sense. For this, a clean single breaking change as of now  would be to match the managed nativeaot and mono results. Does this seem acceptable? If so, I will make that change. Do you think we can get this into .NET 8 if we try to wrap this up this up in the coming week?

Reply:
If we are going to change the behavior, I think we should change it to the saturating behavior that we want to standardize on.

It is unnecessarily complicated from customer point of view to go from the current behavior to the behavior proposed in this PR in .NET 8, and then from that behavior to the desired saturating behavior in .NET 9.

Reply:
So to clarify, you are asking if we can make changes to get the saturating behavior for all 3 i.e. JIT, nativeaot and mono. 

And the main behavior under question here is the floating point to integer conversion. Am I right?

Reply:
Right, I think it would be best to go to the saturating behavior, if we are doing any behavior changes.

As @tannergooding said, switching to the saturating behavior on all runtime flavors is more involved change (see #82781 for WIP version of the change).

Reply:
@tannergooding @jkotas can we get some consensus on this then? As I understand from reading the comments about upgrading to saturating behavior, that requires more work that likely can't be done in the .NET 8 release timeframe.

Does that mean this PR is blocked?

Reply:
Thanks @jkotas 
@tannergooding @jkotas I just skimmed through the changes in https://github.com/dotnet/runtime/pull/82781. At this point we could do 2 things, 

1. Only optimize the 2 cases in this PR for AVX512
        a. ulong -> double
        b. ulong -> float

and take up the following 2 cases in a separate PR which can pick up some changes from https://github.com/dotnet/runtime/pull/82781 
        a. double -> ulong
        b. float -> ulong

This PR was about enabling AVX512 for existing casts and we did not plan this big changes for modifying existing behavior to a saturating behavior.

2. Pickup changes from https://github.com/dotnet/runtime/pull/82781 in this PR and try to optimize all the cases here. i.e.
        double/float <-> ulong

I am more inclined towards 1. since this PR was about enabling AVX512 intrinsic in some of the existing casts and not about modifying existing behavior to a saturating behavior.

**What would be a good way to proceed?**

Reply:
> Only optimize the 2 cases in this PR for AVX512
> a. ulong -> double
> b. ulong -> float

This sounds good to me. There should not be any observable behavior changes introduced by this optimization.

Reply:
Hi folks,

I've gotten a little lost following some of the conversation around the performance numbers (which we are gathering again and checking over) and the saturation/sentinel current/future changes, so I'm just looking for some clarifications.

It sounded like we want to isolate the saturation changes to its own PR as that is a bigger set of changes, and that even with some change of behavior between this PR and existing conversions, it still fixes an oddity on x64/x86 and allows for a potential performance optimization.

Are we saying that we will only do the two `ulong -> double` and `ulong -> float` cases now and address `double -> ulong` and `float -> ulong` after #82781? @tannergooding @jkotas

Reply:
> Are we saying that we will only do the two ulong -> double and ulong -> float cases now and address double -> ulong and float -> ulong after https://github.com/dotnet/runtime/pull/82781?

Yes, I think that this is a good plan.